src/databricks/labs/dqx/check_funcs.py

```diff
@@ -1623,20 +1753,83 @@ def _convert_ipv4_to_bits(ip_col: Column) -> Column:
     return _extract_octets_to_bits(ip_col, DQPattern.IPV4_ADDRESS.value)
 
 
-def _convert_cidr_to_bits_and_prefix(cidr_col: Column) -> tuple[Column, Column]:
+def _convert_ipv4_cidr_to_bits_and_prefix(cidr_col: Column) -> tuple[Column, Column]:
     """Returns binary IP and prefix length from CIDR (e.g., '192.168.1.0/24')."""
     ip_bits = _extract_octets_to_bits(cidr_col, DQPattern.IPV4_CIDR_BLOCK.value)
     # The 5th capture group in the regex pattern corresponds to the CIDR prefix length.
     prefix_length = F.regexp_extract(cidr_col, DQPattern.IPV4_CIDR_BLOCK.value, 5).cast("int").alias("prefix_length")
     return ip_bits, prefix_length
 
 
-def _get_network_address(ip_bits: Column, prefix_length: Column) -> Column:
+def _get_network_address(ip_bits: Column, prefix_length: Column, total_bits: int) -> Column:
     """
     Returns the network address from IP bits using the CIDR prefix length.
 
-    :param ip_bits: 32-bit binary string representation of the IPv4 address
+    :param ip_bits: Binary string representation of the IP address (32 or 128 bits)
     :param prefix_length: Prefix length for CIDR notation
-    :return: Network address as a 32-bit binary string
+    :param total_bits: Total number of bits in the IP address (32 for IPv4, 128 for IPv6)
+    :return: Network address as a binary string of the same total length
+    """
+    return F.rpad(F.substring(ip_bits, 1, prefix_length), total_bits, "0")
+
+
+def _get_normalized_ipv6_hextets(ip_col: Column) -> Column:
+    """
+    Returns a normalized IPv6 as an array of 8 padded hextets.
+    Option 1
+    Example: '::1' -> '00000000...0001']
+    Option 2
+    Example: '::1' -> ['0000', '0000', ..., '0001']
     """
-    return F.rpad(F.substring(ip_bits, 1, prefix_length), 32, "0")
+    parts = F.split(ip_col, "::")
+    is_compressed = F.size(parts) == 2
+
+    left_hextets = F.array_remove(F.split(parts.getItem(0), ":"), "")
+    right_hextets = F.array_remove(F.split(parts.getItem(1), ":"), "")
+
+    num_zeros_needed = F.lit(IPV6_MAX_HEXTET_COUNT) - (F.size(left_hextets) + F.size(right_hextets))
+    zeros = F.array_repeat(F.lit("0000"), num_zeros_needed)
+
+    unpadded_array = F.when(is_compressed, F.concat(left_hextets, zeros, right_hextets)).otherwise(F.split(ip_col, ":"))
+
+    # Option 1
+    return F.array_join(unpadded_array, ":").alias("ip_bits")
+
+    # Option 2
+    # return F.transform(
+    #     unpadded_array,
+    #     lambda hextet: F.lpad(hextet, 4, '0')
```

Comment:
Hi @mwojtyczka, so wanted to get your thoughts on whether it is okay to use higher-order functions in Spark (option 2). Option 1 still achieves the same purpose but I know UDFs are a no-go, but what of higher-order functions?

Reply:
Higher order function should be ok. The computation should remain within the Catalyst optimizer and Spark's JVM execution context. Spark should use native logical plan. There is probably slight serialization overhead due to lambda.

It would be interesting to see what is the performance diff between the option 1 and 2.